metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/sync/ViewSynchronizer.scala

```diff
@@ -0,0 +1,151 @@
+          aggregateStatus(statuses).pure[F]
+
+        case _ =>
+          // We traced all responses, so we can detect if we're in an endless loop.
+          Timer[F].sleep(retryTimeout) >> sync
+      }
+  }
+
+  private def getAndValidateStatus(
+      from: A#PKey
+  ): F[Option[Validated[Status[A]]]] =
+    getStatus(from).flatMap {
+      case None =>
+        none.pure[F]
+
+      case Some(status) =>
+        validate(from, status) match {
+          case Left((error, hint)) =>
+            tracers.invalidStatus(status, error, hint).as(none)
+          case Right(valid) =>
+            valid.some.pure[F]
+        }
+    }
+
+  private def validate(
+      from: A#PKey,
+      status: Status[A]
+  ): Either[
+    (ProtocolError.InvalidQuorumCertificate[A], ViewSynchronizer.Hint),
+    Validated[Status[A]]
+  ] =
+    for {
+      _ <- validateQC(from, status.prepareQC)(
+        checkPhase(Phase.Prepare),
+        checkSignature,
+        checkVisible(status),
+        qc =>
+          check(
+            qc.viewNumber >= status.commitQC.viewNumber,
+            "Prepare Q.C. lower than Commit Q.C."
+          )
+      )
+      _ <- validateQC(from, status.commitQC)(
+        checkPhase(Phase.Commit),
+        checkSignature,
+        checkVisible(status)
+      )
+    } yield Validated[Status[A]](status)
+
+  private def check(cond: Boolean, hint: => String) =
+    if (cond) none else hint.some
+
+  private def checkPhase(phase: Phase)(qc: QuorumCertificate[A]) =
+    check(phase == qc.phase, s"Phase should be $phase.")
+
+  private def checkSignature(qc: QuorumCertificate[A]) =
+    check(Signing[A].validate(federation, qc), "Invalid signature.")
+
+  private def checkVisible(status: Status[A])(qc: QuorumCertificate[A]) =
+    check(
+      status.viewNumber >= qc.viewNumber,
+      "View number of status earlier than Q.C."
+    )
+
+  private def validateQC(from: A#PKey, qc: QuorumCertificate[A])(
+      checks: (QuorumCertificate[A] => Option[String])*
+  ) =
+    checks.toList.traverse { check =>
+      check(qc)
+        .map { hint =>
+          ProtocolError.InvalidQuorumCertificate(from, qc) -> hint
+        }
+        .toLeft(())
+    }
+}
+
+object ViewSynchronizer {
+
+  /** Extra textual description for errors. */
+  type Hint = String
+
+  /** Send a network request to get the status of a replica. */
+  type GetStatus[F[_], A <: Agreement] = A#PKey => F[Option[Status[A]]]
+
+  def aggregateStatus[A <: Agreement](
+      statuses: NonEmptySeq[Status[A]]
+  ): Status[A] = {
+    val prepareQC = statuses.map(_.prepareQC).maximumBy(_.viewNumber)
+    val commitQC  = statuses.map(_.commitQC).maximumBy(_.viewNumber)
+    val viewNumber =
+      math.max(median(statuses.map(_.viewNumber)), prepareQC.viewNumber)
+    Status(
+      viewNumber = ViewNumber(viewNumber),
+      prepareQC = prepareQC,
+      commitQC = commitQC
+    )
+  }
+
+  def median[T: Order](xs: NonEmptySeq[T]): T =
```

Comment:
This is not technically a median, as `median([0, 2]) === 1`. I think this fits the definition of a [medoid](https://en.wikipedia.org/wiki/Medoid). I'm fine with this, but we should perhaps make it clear in a comment?

Reply:
What if used a _mode_ instead? We could calculate all modes and take a "median" of those, if in an extreme case all compared view number are different.

Reply:
Thanks, I didn't know about medoids :)

I named _median_ this because I thought this is the most intuitive name that people will immediately have a pretty good understand of what we are doing and why. I know it doesn't match the traditional definition, I was thinking about it more as an [ordinal value](https://stats.stackexchange.com/questions/11219/median-value-on-ordinal-scales). I thought taking the average of the middle numbers would land you in a no-mans-land, rather than on a view number occupied by at least somebody. You are right, I'll make a comment about it.

No idea whether _mode_ would be better. If as you say the federation is spread out over many view numbers, adversarial nodes could team up to present random maximums to different honest parties, keeping everyone scattered. Like if Alice, Bob and Charlie are honest and have different view numbers, say, 10, 20 and 30, Eve and Edward could tell Alice that they both have 40, to Bob that they are both 50, to Charlie they are both 60. 

Are you're suggesting then that all the others' single-member modes would be mixed in with the modes, so one would pick from [10,20,30,50], the other from [10,20,30,60]? How do you take the height of the mode into account in this case? 

I guess a median is also susceptible to such attack when the network is completely out of sync. Mode in that case could be better if there's at least `f+1` people in the same view.

Reply:
Sorry, I meant categorical values, not ordinal.

Reply:
Yeah, it wasn't a well thought-out suggestion. Both approaches could be attacked as you say.

But I'm thinking of different approach now. What if we sync to at most `max(prepareQC.viewNumber, commitQC.viewNumber + 1)` or maybe just `lastQC.viewNumber + 1`? This could require an honest node to roll back its view number, but only in the case that the preceding rounds were not successful at producing a QC. I realise that this would affect leader selection, but once the honest nodes are in sync, they could continue even if the selected leader is dishonest. Would that make sense?

I'm not sure, seems risky. Looking for another approach as it seems to me any statistical method could be susceptible to attack.

Reply:
Actually I am doing a [+1](https://github.com/input-output-hk/metronome/blob/pm-3323-example-app-integ-test/metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/ConsensusService.scala#L222) now, trying to front run the view I hope everyone else will be finishing soon.

Reply:
Maybe it's worth creating a separate ticket to do some research on this topic. I remember reading that Libra does regular gossip about status, but that changes the complexity of the protocol. We can find out what they do exactly to better understand the landscape.

Reply:
Yes, I agree about the separate ticket. This definitely warrants more research, but it's unlikely to cause issues for now, unless more than 1 actors collude to disrupt syncing.

Reply:
Cool, created https://jira.iohk.io/browse/PM-3346